File: 0-mock-server/master-ws-server/src/DeviceConnectionManager.ts
import WebSocket from 'ws';
import { DeviceType, DeviceInfo, DeviceRegistration } from './types';

/**
 * 设备连接管理器
 * 管理设备注册token和连接状态
 */
export class DeviceConnectionManager {
  /** 待连接设备映射: token -> DeviceInfo */
  private pendingDevices = new Map<string, DeviceInfo>();

  /** Master设备映射: deviceName -> {socket, info, lastHeartbeat} */
  private masterDevices = new Map<string, { socket: WebSocket; info: DeviceInfo; lastHeartbeat: number }>();

  /** Slave设备映射: deviceName -> {socket, info, lastHeartbeat} */
  private slaveDevices = new Map<string, { socket: WebSocket; info: DeviceInfo; lastHeartbeat: number }>();

  /** Master设备ID到设备名称的映射 */
  private masterIdToName = new Map<string, string>();

  /** Master设备名称到关联的slave设备名称列表的映射 */
  private masterToSlaves = new Map<string, Set<string>>();

  /**
   * 预注册设备(HTTP注册阶段)
   * 返回token用于后续WebSocket连接
   */
  preRegisterDevice(registration: DeviceRegistration, token: string): { success: boolean; error?: string } {
    const { type, deviceId, deviceName, masterDeviceId } = registration;

    // 检查设备名称是否已存在
    if (this.masterDevices.has(deviceName) || this.slaveDevices.has(deviceName)) {
      return { success: false, error: 'Device name already exists' };
    }

    // 检查是否已经有相同名称的待注册设备
    for (const pendingInfo of this.pendingDevices.values()) {
      if (pendingInfo.deviceName === deviceName) {
        return { success: false, error: 'Device name already in pending registration' };
      }
    }

    // 如果是slave设备,检查master是否已连接
    if (type === DeviceType.SLAVE) {
      if (!masterDeviceId) {
        return { success: false, error: 'masterDeviceId is required for slave device' };
      }

      const masterName = this.masterIdToName.get(masterDeviceId);
      if (!masterName) {
        return { success: false, error: 'Master device not connected' };
      }
    }

    const deviceInfo: DeviceInfo = {
      ...registration,
      connectedAt: new Date(),
      token
    };

    this.pendingDevices.set(token, deviceInfo);

    console.log(`[预注册] ${type === DeviceType.MASTER ? 'Master' : 'Slave'}: ${deviceName} (Token: ${token.substring(0, 8)}...)`);

    return { success: true };
  }

  /**
   * 通过token连接设备(WebSocket连接阶段)
   */
  connectDeviceWithToken(socket: WebSocket, token: string): { success: boolean; error?: string; deviceInfo?: DeviceInfo } {
    const deviceInfo = this.pendingDevices.get(token);

    if (!deviceInfo) {
      return { success: false, error: 'Invalid or expired token' };
    }

    const { type, deviceName, deviceId, masterDeviceId } = deviceInfo;

    // 再次检查设备名称是否已被占用
    if (this.masterDevices.has(deviceName) || this.slaveDevices.has(deviceName)) {
      this.pendingDevices.delete(token);
      return { success: false, error: 'Device name already connected' };
    }

    // 如果是slave设备,再次检查master是否还在线
    if (type === DeviceType.SLAVE && masterDeviceId) {
      const masterName = this.masterIdToName.get(masterDeviceId);
      if (!masterName || !this.masterDevices.has(masterName)) {
        this.pendingDevices.delete(token);
        return { success: false, error: 'Master device disconnected' };
      }
    }

    // 移除待注册状态
    this.pendingDevices.delete(token);

    // 更新连接时间
    deviceInfo.connectedAt = new Date();

    // 添加到已连接设备
    if (type === DeviceType.MASTER) {
      this.masterDevices.set(deviceName, { socket, info: deviceInfo, lastHeartbeat: Date.now() });
      this.masterIdToName.set(deviceId, deviceName);
      this.masterToSlaves.set(deviceName, new Set());
      console.log(`[Master] 连接成功: ${deviceName} (ID: ${deviceId})`);
    } else {
      this.slaveDevices.set(deviceName, { socket, info: deviceInfo, lastHeartbeat: Date.now() });
      const masterName = this.masterIdToName.get(masterDeviceId!);
      if (masterName) {
        this.masterToSlaves.get(masterName)?.add(deviceName);
        console.log(`[Slave] 连接成功: ${deviceName} (ID: ${deviceId}) -> Master: ${masterName}`);
      }
    }

    return { success: true, deviceInfo };
  }

  /**
   * 清理过期的待注册设备(超过5分钟)
   */
  cleanExpiredPendingDevices() {
    const now = Date.now();
    const expireTime = 5 * 60 * 1000; // 5分钟

    for (const [token, info] of this.pendingDevices.entries()) {
      if (now - info.connectedAt.getTime() > expireTime) {
        this.pendingDevices.delete(token);
        console.log(`[清理] 过期的待注册设备: ${info.deviceName}`);
      }
    }
  }

  /**
   * 获取master设备连接
   */
  getMasterDevice(deviceName: string) {
    return this.masterDevices.get(deviceName);
  }

  /**
   * 通过master设备ID获取master设备连接
   */
  getMasterDeviceById(deviceId: string) {
    const deviceName = this.masterIdToName.get(deviceId);
    return deviceName ? this.masterDevices.get(deviceName) : undefined;
  }

  /**
   * 获取slave设备连接
   */
  getSlaveDevice(deviceName: string) {
    return this.slaveDevices.get(deviceName);
  }

  /**
   * 获取master设备的所有关联slave设备
   */
  getSlavesByMaster(masterName: string): Array<{ socket: WebSocket; info: DeviceInfo; lastHeartbeat: number }> {
    const slaveNames = this.masterToSlaves.get(masterName);
    if (!slaveNames) {
      return [];
    }

    const slaves: Array<{ socket: WebSocket; info: DeviceInfo; lastHeartbeat: number }> = [];
    slaveNames.forEach(slaveName => {
      const slave = this.slaveDevices.get(slaveName);
      if (slave) {
        slaves.push(slave);
      }
    });

    return slaves;
  }

  /**
   * 获取slave设备关联的master设备
   */
  getMasterBySlaveInfo(slaveInfo: DeviceInfo) {
    if (!slaveInfo.masterDeviceId) {
      return undefined;
    }
    return this.getMasterDeviceById(slaveInfo.masterDeviceId);
  }

  /**
   * 通过设备名称查找指定的slave设备
   * 并验证该slave是否属于指定的master
   */
  getSlaveByNameAndMaster(slaveDeviceName: string, masterName: string) {
    const slave = this.slaveDevices.get(slaveDeviceName);
    if (!slave) {
      return undefined;
    }

    // 验证该slave是否属于该master
    const slaveNames = this.masterToSlaves.get(masterName);
    if (!slaveNames || !slaveNames.has(slaveDeviceName)) {
      return undefined;
    }

    return slave;
  }

  /**
   * 断开master设备连接
   * 会自动断开所有关联的slave设备
   */
  disconnectMaster(deviceName: string) {
    const master = this.masterDevices.get(deviceName);
    if (!master) {
      return;
    }

    const slaveNames = this.masterToSlaves.get(deviceName);
    if (slaveNames) {
      slaveNames.forEach(slaveName => {
        const slave = this.slaveDevices.get(slaveName);
        if (slave) {
          console.log(`[Slave] 因master断开而自动断开: ${slaveName}`);
          slave.socket.close();
          this.slaveDevices.delete(slaveName);
        }
      });
      this.masterToSlaves.delete(deviceName);
    }

    this.masterIdToName.delete(master.info.deviceId);
    this.masterDevices.delete(deviceName);
    console.log(`[Master] 断开连接: ${deviceName}`);
  }

  /**
   * 断开slave设备连接
   */
  disconnectSlave(deviceName: string) {
    const slave = this.slaveDevices.get(deviceName);
    if (!slave) {
      return;
    }

    const { masterDeviceId } = slave.info;
    if (masterDeviceId) {
      const masterName = this.masterIdToName.get(masterDeviceId);
      if (masterName) {
        this.masterToSlaves.get(masterName)?.delete(deviceName);
      }
    }

    this.slaveDevices.delete(deviceName);
    console.log(`[Slave] 断开连接: ${deviceName}`);
  }

  /**
   * 通过socket查找设备信息
   */
  findDeviceBySocket(socket: WebSocket): { type: DeviceType; name: string; info: DeviceInfo } | undefined {
    for (const [name, device] of this.masterDevices.entries()) {
      if (device.socket === socket) {
        return { type: DeviceType.MASTER, name, info: device.info };
      }
    }

    for (const [name, device] of this.slaveDevices.entries()) {
      if (device.socket === socket) {
        return { type: DeviceType.SLAVE, name, info: device.info };
      }
    }

    return undefined;
  }

  /**
   * 获取统计信息
   */
  getStats() {
    return {
      masterCount: this.masterDevices.size,
      slaveCount: this.slaveDevices.size,
      pendingCount: this.pendingDevices.size,
      masters: Array.from(this.masterDevices.entries()).map(([name, device]) => ({
        name,
        deviceId: device.info.deviceId,
        connectedAt: device.info.connectedAt,
        slaveCount: this.masterToSlaves.get(name)?.size || 0
      })),
      slaves: Array.from(this.slaveDevices.entries()).map(([name, device]) => ({
        name,
        deviceId: device.info.deviceId,
        masterDeviceId: device.info.masterDeviceId,
        connectedAt: device.info.connectedAt
      }))
    };
  }

  /**
   * 更新设备心跳时间
   */
  updateHeartbeat(deviceName: string, deviceType: DeviceType) {
    if (deviceType === DeviceType.MASTER) {
      const device = this.masterDevices.get(deviceName);
      if (device) {
        device.lastHeartbeat = Date.now();
      }
    } else {
      const device = this.slaveDevices.get(deviceName);
      if (device) {
        device.lastHeartbeat = Date.now();
      }
    }
  }

  /**
   * 检查并断开超时的设备
   * @returns 被断开的设备列表
   */
  checkAndDisconnectTimeoutDevices(timeoutMs: number): Array<{ type: DeviceType; name: string; info: DeviceInfo }> {
    const now = Date.now();
    const disconnectedDevices: Array<{ type: DeviceType; name: string; info: DeviceInfo }> = [];

    // 检查master设备
    for (const [name, device] of this.masterDevices.entries()) {
      if (now - device.lastHeartbeat > timeoutMs) {
        disconnectedDevices.push({ type: DeviceType.MASTER, name, info: device.info });
        device.socket.close();
      }
    }

    // 检查slave设备
    for (const [name, device] of this.slaveDevices.entries()) {
      if (now - device.lastHeartbeat > timeoutMs) {
        disconnectedDevices.push({ type: DeviceType.SLAVE, name, info: device.info });
        device.socket.close();
      }
    }

    return disconnectedDevices;
  }

  /**
   * 获取所有设备列表(用于发送心跳)
   */
  getAllDevices(): Array<{ type: DeviceType; name: string; socket: WebSocket }> {
    const devices: Array<{ type: DeviceType; name: string; socket: WebSocket }> = [];

    for (const [name, device] of this.masterDevices.entries()) {
      devices.push({ type: DeviceType.MASTER, name, socket: device.socket });
    }

    for (const [name, device] of this.slaveDevices.entries()) {
      devices.push({ type: DeviceType.SLAVE, name, socket: device.socket });
    }

    return devices;
  }
}

File: _old_/IMPos2AdapterV1/android/turbomodule-lib/src/main/java/com/impos2/turbomodules/localwebserver/DeviceConnectionManager.kt
package com.impos2.turbomodules.localwebserver

import android.util.Log
import io.ktor.websocket.*
import kotlinx.coroutines.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * 设备连接管理器
 *
 * 负责管理 Master-Slave 设备的注册、连接、消息路由和心跳检测
 *
 * 优化点:
 * 1. 使用 ConcurrentHashMap 确保线程安全
 * 2. 自动清理过期的待注册设备
 * 3. 心跳检测机制,自动断开超时设备
 * 4. Master-Slave 关联管理,Master 断开时自动断开所有 Slave
 */
class DeviceConnectionManager(
    private val config: ServerConfig
) {
    companion object {
        private const val TAG = "DeviceConnectionManager"
        private const val TOKEN_EXPIRE_TIME = 5 * 60 * 1000L // 5分钟
    }

    // 待注册设备 (token -> PendingDevice)
    private val pendingDevices = ConcurrentHashMap<String, PendingDevice>()

    // Master 设备 (deviceName -> DeviceConnection)
    private val masterDevices = ConcurrentHashMap<String, DeviceConnection>()

    // Slave 设备 (deviceName -> DeviceConnection)
    private val slaveDevices = ConcurrentHashMap<String, DeviceConnection>()

    // Master ID 到名称的映射
    private val masterIdToName = ConcurrentHashMap<String, String>()

    // Master 到 Slaves 的映射
    private val masterToSlaves = ConcurrentHashMap<String, MutableSet<String>>()

    // 服务器启动时间
    private val serverStartTime = System.currentTimeMillis()

    // 协程作用域
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    /**
     * 待注册设备信息
     */
    data class PendingDevice(
        val info: DeviceInfo,
        val registrationTime: Long,
        val masterDeviceId: String? = null
    )

    /**
     * 设备连接信息
     */
    data class DeviceConnection(
        val session: WebSocketSession,
        val info: DeviceInfo,
        var lastHeartbeat: Long
    )

    init {
        // 启动定期清理任务
        startCleanupTask()
        // 启动心跳检测任务
        startHeartbeatTask()
    }

    /**
     * 预注册设备，返回 token
     */
    fun preRegisterDevice(registration: DeviceRegistration): RegistrationResponse {
        try {
            // 验证设备名称唯一性
            if (masterDevices.containsKey(registration.deviceName) ||
                slaveDevices.containsKey(registration.deviceName)) {
                return RegistrationResponse(
                    success = false,
                    error = "设备名称已存在: ${registration.deviceName}"
                )
            }

            // 如果是 Slave，验证 Master 是否存在
            if (registration.type == "slave") {
                val masterDeviceId = registration.masterDeviceId
                if (masterDeviceId == null) {
                    return RegistrationResponse(
                        success = false,
                        error = "Slave 设备必须指定 masterDeviceId"
                    )
                }
                if (!masterIdToName.containsKey(masterDeviceId)) {
                    return RegistrationResponse(
                        success = false,
                        error = "Master 设备不存在: $masterDeviceId"
                    )
                }
            }

            // 生成 token
            val token = UUID.randomUUID().toString()
            val deviceInfo = DeviceInfo(
                deviceType = registration.type,
                deviceId = registration.deviceId,
                deviceName = registration.deviceName
            )

            // 保存待注册设备
            pendingDevices[token] = PendingDevice(
                info = deviceInfo,
                registrationTime = System.currentTimeMillis(),
                masterDeviceId = registration.masterDeviceId
            )

            Log.d(TAG, "设备预注册成功: ${registration.deviceName}, token: $token")

            return RegistrationResponse(
                success = true,
                token = token,
                deviceInfo = deviceInfo
            )
        } catch (e: Exception) {
            Log.e(TAG, "预注册设备失败", e)
            return RegistrationResponse(
                success = false,
                error = "预注册失败: ${e.message}"
            )
        }
    }

    /**
     * 通过 token 连接设备
     */
    suspend fun connectDeviceWithToken(token: String, session: WebSocketSession): Boolean {
        val pending = pendingDevices.remove(token) ?: return false

        val connection = DeviceConnection(
            session = session,
            info = pending.info,
            lastHeartbeat = System.currentTimeMillis()
        )

        when (pending.info.deviceType) {
            "master" -> {
                masterDevices[pending.info.deviceName] = connection
                masterIdToName[pending.info.deviceId] = pending.info.deviceName
                // 使用 ConcurrentHashMap.newKeySet() 创建线程安全的 Set
                masterToSlaves[pending.info.deviceName] = ConcurrentHashMap.newKeySet()
                Log.d(TAG, "Master 设备连接成功: ${pending.info.deviceName}")
            }
            "slave" -> {
                slaveDevices[pending.info.deviceName] = connection
                val masterName = masterIdToName[pending.masterDeviceId]
                if (masterName != null) {
                    masterToSlaves[masterName]?.add(pending.info.deviceName)
                    // 通知 Master 有 Slave 连接
                    notifyMasterSlaveConnected(masterName, pending.info)
                }
                Log.d(TAG, "Slave 设备连接成功: ${pending.info.deviceName}")
            }
        }

        return true
    }

    /**
     * 通知 Master 有 Slave 连接
     */
    private suspend fun notifyMasterSlaveConnected(masterName: String, slaveInfo: DeviceInfo) {
        val master = masterDevices[masterName] ?: return

        val notification = buildJsonObject {
            put("from", "system")
            put("id", UUID.randomUUID().toString())
            put("type", SystemMessageTypes.SLAVE_CONNECTED)
            put("data", buildJsonObject {
                put("deviceId", slaveInfo.deviceId)
                put("deviceName", slaveInfo.deviceName)
            })
        }

        try {
            master.session.send(Frame.Text(notification.toString()))
            Log.d(TAG, "已通知 Master: Slave 连接 - ${slaveInfo.deviceName}")
        } catch (e: Exception) {
            Log.e(TAG, "通知 Master 失败", e)
        }
    }

    /**
     * 路由消息
     */
    suspend fun routeMessage(message: MessageWrapper, fromDevice: String) {
        // 判断发送者类型
        val isMaster = masterDevices.containsKey(fromDevice)

        if (isMaster) {
            // Master 发送消息
            routeMessageFromMaster(message, fromDevice)
        } else {
            // Slave 发送消息
            routeMessageFromSlave(message, fromDevice)
        }
    }

    /**
     * 路由来自 Master 的消息
     */
    private suspend fun routeMessageFromMaster(message: MessageWrapper, masterName: String) {
        val targetDevice = message.targetDevice

        if (targetDevice != null) {
            // 发送给指定 Slave
            val slave = slaveDevices[targetDevice]
            if (slave != null) {
                try {
                    slave.session.send(Frame.Text(Json.encodeToString(message)))
                    Log.d(TAG, "消息已发送: Master -> $targetDevice")
                } catch (e: Exception) {
                    Log.e(TAG, "发送消息失败", e)
                }
            }
        } else {
            // 广播给所有关联的 Slave
            val slaves = masterToSlaves[masterName] ?: emptySet()
            slaves.forEach { slaveName ->
                val slave = slaveDevices[slaveName]
                if (slave != null) {
                    try {
                        slave.session.send(Frame.Text(Json.encodeToString(message)))
                    } catch (e: Exception) {
                        Log.e(TAG, "广播消息失败: $slaveName", e)
                    }
                }
            }
            Log.d(TAG, "消息已广播: Master -> ${slaves.size} Slaves")
        }
    }

    /**
     * 路由来自 Slave 的消息
     */
    private suspend fun routeMessageFromSlave(message: MessageWrapper, slaveName: String) {
        // 查找 Slave 关联的 Master
        val masterName = masterToSlaves.entries.find { it.value.contains(slaveName) }?.key
        if (masterName != null) {
            val master = masterDevices[masterName]
            if (master != null) {
                try {
                    master.session.send(Frame.Text(Json.encodeToString(message)))
                    Log.d(TAG, "消息已发送: Slave($slaveName) -> Master($masterName)")
                } catch (e: Exception) {
                    Log.e(TAG, "发送消息失败", e)
                }
            }
        }
    }

    /**
     * 更新设备心跳时间
     */
    fun updateHeartbeat(deviceName: String) {
        masterDevices[deviceName]?.lastHeartbeat = System.currentTimeMillis()
        slaveDevices[deviceName]?.lastHeartbeat = System.currentTimeMillis()
    }

    /**
     * 通过 session 获取设备名称
     */
    fun getDeviceNameBySession(session: WebSocketSession): String? {
        masterDevices.entries.find { it.value.session == session }?.let {
            return it.key
        }
        slaveDevices.entries.find { it.value.session == session }?.let {
            return it.key
        }
        return null
    }

    /**
     * 断开设备
     */
    suspend fun disconnectDevice(deviceName: String) {
        val master = masterDevices.remove(deviceName)
        if (master != null) {
            // 断开 Master，同时断开所有关联的 Slave
            val slaves = masterToSlaves.remove(deviceName) ?: emptySet()
            slaves.forEach { slaveName ->
                slaveDevices.remove(slaveName)
            }
            masterIdToName.entries.removeIf { it.value == deviceName }
            Log.d(TAG, "Master 设备已断开: $deviceName, 关联 Slave: ${slaves.size}")
            return
        }

        val slave = slaveDevices.remove(deviceName)
        if (slave != null) {
            // 断开 Slave，通知 Master
            val masterName = masterToSlaves.entries.find { it.value.contains(deviceName) }?.key
            if (masterName != null) {
                masterToSlaves[masterName]?.remove(deviceName)
                notifyMasterSlaveDisconnected(masterName, slave.info)
            }
            Log.d(TAG, "Slave 设备已断开: $deviceName")
        }
    }

    /**
     * 通知 Master 有 Slave 断开
     */
    private suspend fun notifyMasterSlaveDisconnected(masterName: String, slaveInfo: DeviceInfo) {
        val master = masterDevices[masterName] ?: return

        val notification = buildJsonObject {
            put("from", "system")
            put("id", UUID.randomUUID().toString())
            put("type", SystemMessageTypes.SLAVE_DISCONNECTED)
            put("data", buildJsonObject {
                put("deviceId", slaveInfo.deviceId)
                put("deviceName", slaveInfo.deviceName)
            })
        }

        try {
            master.session.send(Frame.Text(notification.toString()))
            Log.d(TAG, "已通知 Master: Slave 断开 - ${slaveInfo.deviceName}")
        } catch (e: Exception) {
            Log.e(TAG, "通知 Master 失败", e)
        }
    }

    /**
     * 启动定期清理任务
     */
    private fun startCleanupTask() {
        scope.launch {
            while (isActive) {
                delay(60000) // 每分钟清理一次
                cleanExpiredPendingDevices()
            }
        }
    }

    /**
     * 清理过期的待注册设备
     */
    private fun cleanExpiredPendingDevices() {
        val now = System.currentTimeMillis()
        val expired = pendingDevices.filter { (_, pending) ->
            now - pending.registrationTime > TOKEN_EXPIRE_TIME
        }
        expired.keys.forEach { token ->
            pendingDevices.remove(token)
        }
        if (expired.isNotEmpty()) {
            Log.d(TAG, "清理过期待注册设备: ${expired.size} 个")
        }
    }

    /**
     * 启动心跳检测任务
     */
    private fun startHeartbeatTask() {
        scope.launch {
            while (isActive) {
                delay(config.heartbeatInterval)
                sendHeartbeat()
                checkTimeoutDevices()
            }
        }
    }

    /**
     * 发送心跳消息
     */
    private suspend fun sendHeartbeat() {
        val heartbeat = buildJsonObject {
            put("from", "system")
            put("id", UUID.randomUUID().toString())
            put("type", SystemMessageTypes.HEARTBEAT)
            put("data", buildJsonObject {})
        }

        val allDevices = masterDevices.values + slaveDevices.values
        allDevices.forEach { device ->
            try {
                device.session.send(Frame.Text(heartbeat.toString()))
            } catch (e: Exception) {
                Log.e(TAG, "发送心跳失败", e)
            }
        }
    }

    /**
     * 检查超时设备
     */
    private suspend fun checkTimeoutDevices() {
        val now = System.currentTimeMillis()
        val timeout = config.heartbeatTimeout

        val timeoutMasters = masterDevices.filter { (_, device) ->
            now - device.lastHeartbeat > timeout
        }

        val timeoutSlaves = slaveDevices.filter { (_, device) ->
            now - device.lastHeartbeat > timeout
        }

        timeoutMasters.keys.forEach { deviceName ->
            Log.w(TAG, "Master 设备超时: $deviceName")
            disconnectDevice(deviceName)
        }

        timeoutSlaves.keys.forEach { deviceName ->
            Log.w(TAG, "Slave 设备超时: $deviceName")
            disconnectDevice(deviceName)
        }
    }

    /**
     * 获取统计信息
     */
    fun getStats(): ServerStats {
        return ServerStats(
            masterCount = masterDevices.size,
            slaveCount = slaveDevices.size,
            pendingCount = pendingDevices.size,
            uptime = System.currentTimeMillis() - serverStartTime
        )
    }

    /**
     * 关闭管理器
     */
    fun close() {
        try {
            // 使用 runBlocking 来调用 suspend 函数
            runBlocking {
                // 断开所有 Master 设备
                masterDevices.keys.toList().forEach { deviceName ->
                    try {
                        disconnectDevice(deviceName)
                    } catch (e: Exception) {
                        Log.e(TAG, "断开 Master 设备失败: $deviceName", e)
                    }
                }

                // 断开所有 Slave 设备
                slaveDevices.keys.toList().forEach { deviceName ->
                    try {
                        disconnectDevice(deviceName)
                    } catch (e: Exception) {
                        Log.e(TAG, "断开 Slave 设备失败: $deviceName", e)
                    }
                }
            }

            // 清理所有数据
            pendingDevices.clear()
            masterDevices.clear()
            slaveDevices.clear()
            masterIdToName.clear()
            masterToSlaves.clear()

            // 取消协程作用域
            scope.cancel()

            Log.d(TAG, "DeviceConnectionManager 已关闭")
        } catch (e: Exception) {
            Log.e(TAG, "关闭 DeviceConnectionManager 失败", e)
        }
    }
}
